stubs/PyYAML/yaml/__init__.pyi

```diff
@@ -258,12 +261,32 @@ def safe_dump(
     tags=...,
     sort_keys: bool = ...,
 ) -> _Yaml: ...
-def add_implicit_resolver(tag, regexp, first=..., Loader=..., Dumper=...): ...
-def add_path_resolver(tag, path, kind=..., Loader=..., Dumper=...): ...
-def add_constructor(tag: _Str, constructor: Callable[[Loader, Node], Any], Loader: Loader = ...): ...
-def add_multi_constructor(tag_prefix, multi_constructor, Loader=...): ...
-def add_representer(data_type: Type[_T], representer: Callable[[_R, _T], Node], Dumper: Type[_R] = ...) -> None: ...
-def add_multi_representer(data_type: Type[_T], multi_representer: Callable[[_R, _T], Node], Dumper: Type[_R] = ...) -> None: ...
+def add_implicit_resolver(
+    tag: _Str,
+    regexp: Pattern[str],
+    first: Optional[Iterable[Any]] = ...,
+    Loader: Optional[Type[BaseResolver]] = ...,
+    Dumper: Type[BaseResolver] = ...,
+) -> None: ...
+def add_path_resolver(
+    tag: _Str,
+    path: Iterable[Any],
+    kind: Optional[Type[Any]] = ...,
+    Loader: Optional[Type[BaseResolver]] = ...,
+    Dumper: Type[BaseResolver] = ...,
+) -> None: ...
+def add_constructor(
+    tag: _Str, constructor: Callable[[_Constructor, Node], Any], Loader: Optional[Type[_Constructor]] = ...
+) -> None: ...
```

Comment:
If `Loader` is not specified (or is specified as `None`), mypy can't figure out what type `_Constructor` should be. For example, this...

```python
from yaml import add_constructor
add_constructor("lol", (lambda x, y: "wut"))
```

...fails with this error:

```
(env) akuli@akuli-desktop:~/typeshed$ MYPYPATH=stubs/PyYAML mypy foo.py
foo.py:2: error: Cannot infer type argument 1 of "add_constructor"
Found 1 error in 1 file (checked 1 source file)
```

If I read the source code correctly, in this case it can be `Loader`, `FullLoader` or `UnsafeLoader`. To do this, you need two overloads: one when `Loader` is not given or it is None, and one when it is given.

```suggestion
@overload
def add_constructor(
    tag: _Str, constructor: Callable[[Loader | FullLoader | UnsafeLoader, Node], Any], Loader: None = ...
) -> None: ...
@overload
def add_constructor(
    tag: _Str, constructor: Callable[[_Constructor, Node], Any], Loader: Type[_Constructor]
) -> None: ...
```

The same applies to other functions that use `_Constructor` or `_Representer` too.

Reply:
Thanks for the explanation! I addressed them in 8f56493 (+ a small fix in adc766b).

In `add_representer` it's actually a peculiar case that the default of the last parameter is not `None` but `Dumper`, which itself is a subclass of `BaseRepresenter`. So it seems that leaving out this last argument in the first overload is fine:
```python
@overload
def add_representer(data_type: Type[_T], representer: Callable[[Dumper, _T], Node]) -> None: ...
@overload
def add_representer(
    data_type: Type[_T], representer: Callable[[_Representer, _T], Node], Dumper: Type[_Representer] = ...
) -> None: ...
```
because calling `add_representer` with its default third parameter passed explicitly is actually covered by the second overload. (This didn't work for `add_constructor` since the default is `None`, hence the fix in adc766b)

I hope it's fine that way.